src/listen/ldap_sync/active_directory.c

```diff
@@ -0,0 +1,684 @@
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ * @file active_directory.c
+ * @brief LDAP sync callback functions for Active Directory servers.
+ *
+ * @copyright 2022 Network RADIUS SARL (legal@networkradius.com)
+ */
+
+#define LOG_PREFIX "ldap_sync_ad"
+
+#include "active_directory.h"
+#include <freeradius-devel/util/debug.h>
+
+/** Check that a particular attribute is included in the query attribute list
+ *
+ * @param[in] attrs	list to check
+ * @param[in] attr	to look for
+ * @param[in] msg	to include in the warning
+ */
+static void ldap_query_attrs_check(char const **attrs, char const *attr, char const *msg)
+{
+	size_t		len, i;
+
+	if (!attr) return;
+
+	len = talloc_array_length(attrs);
+
+	/*
+	 *	No attributes specified - so all will be returned
+	 */
+	if (len == 0) return;
+
+	for (i = 0; i < len; i++) {
+		if (!attrs[i]) continue;
+		if (strcasecmp(attrs[i], attr) == 0) return;
+		if (strcasecmp(attrs[i], "*") == 0) return;
+	}
+
+	WARN("Attribute %s %s, but not in requested attributes", attr, msg);
+}
+
+/** Terminal characters for parsing LDAP filters
+*/
+static const fr_sbuff_term_t filter_terms = FR_SBUFF_TERMS(
+	L("="),
+	L("<"),
+	L(">"),
+	L("~"),
+	L(":"),
+	L(")"),
+	L("("),
+	L("&"),
+	L("|"),
+	L("!")
+);
+
+/** Parse individual LDAP filter nodes
+ *
+ * @param[in,out] node	to populate with parsed filter
+ * @param[in] sbuff	pointing to filter to parse
+ * @param[in] depth	to indent debug output, indicating nesting of groups
+ *
+ * @return
+ *  - 0 on success
+ *  - < 0 on error
+ *  - 1 if the final advance of the sbuff moved forward.
+ */
+static int ldap_filter_parse_node(ldap_filter_t *node, fr_sbuff_t *sbuff, int depth, char const **attrs)
+{
+	fr_sbuff_marker_t	filter_marker, attr_s_marker, attr_e_marker, value_marker;
+	ldap_filter_t		*child_node;
+
+	/*
+	 *	Firstly, look for the characters which indicate the start of a group of filters
+	 *	to be combined with a logical operator.
+	 */
+	switch(fr_sbuff_current(sbuff)[0]) {
+	case '&':
+		node->logic_op = LDAP_FILTER_LOGIC_AND;
+		node->orig = talloc_asprintf(node, "&");
+		goto sync_filter_new_group;
+
+	case '|':
+		node->logic_op = LDAP_FILTER_LOGIC_OR;
+		node->orig = talloc_asprintf(node, "|");
+		goto sync_filter_new_group;
+
+	case '!':
+		node->logic_op = LDAP_FILTER_LOGIC_NOT;
+		node->orig = talloc_asprintf(node, "!");
+	sync_filter_new_group:
+		if ((fr_sbuff_remaining(sbuff) < 2) || (fr_sbuff_current(sbuff)[1] != '(')) {
```

Comment:
Use `fr_sbuff_extend` instead of remaining.   Add an `fr_sbuff_peek(_sbuff, offset)` which returns '\0' if the access is out of bounds if you need an offset.

Reply:
Isn't `fr_sbuff_extend` about extending buffers you are writing to?  I'm after seeing how close we are to the end of a buffer I'm reading from

Reply:
It works for both depending on what the underlying sbuff is.  The `fr_sbuff_t` can be initialised with different extend functions.  Where we're using it for printing the extend function can realloc the underlying buffer for more space, or potentially write data in the buffer to a socket, and shift everything to the left.  When we're using it for parsing, we're usually shifting parsed stuff to the left, and copying in more data at the end of the buffer from a `FILE *` or another data source.

The point of markers isn't really to delimit sections, you use recursive function calls for that.  Markers are used if you might need to refer back to text earlier in the buffer, or write to a position earlier in the buffer.  They're used to stop the data you're referring to being shifted out of the buffer.